DefaultedRegistryValue, Taggable & StringRepresentable expansion

MrHell228 opened the pull request:
**SpongeAPI** | [Sponge](https://github.com/SpongePowered/Sponge/pull/4198)

QOL PR with big amount of lines but very small amont of actual changes.

List of changes:
- Currently `Taggable` has `registryType` method, and I think it would be nice to have it for `DefaultedRegistryValue` too to be able to retrieve registry key just from value without need to pass registry. So old `DefaultedRegistryValue` methods moved to `RegisryValue` interface (i guess it makes sense).
- `#tags` & `#is` in `Taggable` now defaulted so interface would be more friendly for plugin devs to use.
- All `DefaultedRegistryValue` and `Taggable` subtypes now explicitly implement `#registryType`.
- All vanilla registry types (except `WorldType` & `WorldArchetypeType`) implement `Taggable` because minecraft allows adding tags for them with data packs
- A lot of registry types implement `DefaultedRegistryValue` because they were missing it
- A lot of registry types implement `StringRepresentable` because they were missing it
- Added missing builtin tag groups

If override `RegistryValue` methods in `DefaultedRegistryValue`, this whole PR shouldn't cause even binary breaking change. Don't know if it's needed though.

MrHell228 committed:
Introduce RegistryValue & DefaultedTaggable, move some things around

MrHell228 committed:
update old Taggable types

MrHell228 committed:
make vanilla registry values Taggable

MrHell228 committed:
update DefaultRegistryValue types

MrHell228 committed:
Add missing Taggable types

MrHell228 committed:
Add missing tags

MrHell228 committed:
Add missing DefaultedRegistryValue & StringRepresentable types

MrHell228 committed:
remove Taggable, rename DefaultedTaggable to Taggable

MrHell228 committed:
spotlessApply

Zidane commented on src/main/java/org/spongepowered/api/tag/Taggable.java:
```diff
@@ -24,37 +24,35 @@
 
 /**
- * A type that may be included in one or more {@link Tag} collections.
+ * A {@link DefaultedRegistryValue} that may be included in one or more {@link Tag} collections.
```
This feels like implementation creep onto the API where taggables must be defaulted registry values.

MrHell228 replied:
ah, I forgot to revert this line, sorry. (`Taggable` interface extends `DefaultedRegsistryValue` even without this PR though)

Zidane commented on src/main/java/org/spongepowered/api/registry/RegistryValue.java:
```diff
@@ -0,0 +1,50 @@
+ * within a {@link DefaultedRegistryType defaulted registry}.
+ */
+@SuppressWarnings("unchecked")
+public interface RegistryValue<T extends RegistryValue<T>> {
```
Give me "What is an example of why Java needs extension methods for $200 Alex".

In all seriousness, I detest stuff like this. Since we can't fix the JDK, maybe this needs a better name. Because not all things that are RegistryValues can extend this...

MrHell228 replied:
Maybe something like `KnownRegistryValue` or `RegistryValueCandidate`?

And honestly, I don't know if these "old" methods are even needed with `DefaultedRegistryValue` being enhanced. I guess having 2 registries for the same type is unnecessary because the same goal can be achieveved with `Tag`s

MrHell228 committed:
nuke RegistryValue, add overloads that take RegistrtHolder

MrHell228 commented:
Nuked `RegistryValue` interface.
Added overloads that take `RegistryHolder` for methods in `DefaultedRegistryValue` and `Taggable`.
Changed return type of `Taggable#tags` from `Collection` to `Stream`

MrHell228 committed:
abstract Taggable

MrHell228 committed:
Revert "abstract Taggable"

This reverts commit e965ac0877399870de8cf62133a44efe7e26a75e.

MrHell228 committed:
move #registryType a bit
